src/Processors/Sources/RemoteSource.cpp

```diff
@@ -72,19 +72,33 @@ ISource::Status RemoteSource::prepare()
     if (is_async_state)
         return Status::Async;
 
+    if (executor_finished)
+        return Status::Finished;
+
     Status status = ISource::prepare();
     /// To avoid resetting the connection (because of "unfinished" query) in the
     /// RemoteQueryExecutor it should be finished explicitly.
     if (status == Status::Finished)
     {
-        query_executor->finish();
         is_async_state = false;
-        return status;
+        need_drain = true;
+        return Status::Ready;
     }
 
     return status;
 }
 
+void RemoteSource::work()
```

Comment:
probably add some comment here why move drain to work from prepare, to facilitate others' understanding.

Reply:
okay